Add Mux and Demux to hydro

Squelsh opened the pull request:
I was missing a Demux tool. So I brought the old code back to life.
While working on that, I also fixed the service-servers of the Mux tool, so more than one Mux can be used.

Code is tested for hydro. But I think it can also be merged into newer versions. As I have no newer version running, I could not test. ==> Would be nice, if someone could do that.

If my request gets merged, I am happy to update the ROS WIKI page.

Commit:
Moved ServiceServers to private node handle.
That way more than one mux can coexist.

Commit:
Fixed the old Demux code and brought it back alive.
Mainly followed the code from the mux.
Added according service descriptions.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

tfoote commented on tools/topic_tools/src/demux.cpp:
```diff
@@ -1,9 +1,9 @@
+// to 1 of N output topics. A service is provided to select between the outputs
 //
-// Copyright (C) 2009, Morgan Quigley
+// Copyright (C) 2014, Andreas Hermann
```
The copyright should be side by side not replace, even though it's mostly new. 

```
Copyright (C) 2009, Morgan Quigley
Copyright (C) 2014, Andreas Hermann
```

dirk-thomas commented:
This comment (https://github.com/Squelsh/ros_comm/blob/hydro-devel/tools/topic_tools/src/mux.cpp#L301) states explicitly that the services are provided on the node "mux" and suggests using remapping in order to use multiple mux instances.

This patch would effectively change the public API of the tool which is not a feasible change for Hydro. Even for Indigo it is questionable if that should be changed.

The demux part is definitely a good additional. Thank you for porting this functionality.

Squelsh commented:
Hi Tully, hi Dirk,
sorry for the quirks. I totally agree with your comments.
@tfoote : I corrected the Copyright as suggested.
@dirk-thomas : I reverted my changes and adapted the demux to behave the same way as the Mux.

Commit:
Revert "Moved ServiceServers to private node handle."

This reverts commit b958d21ffd614bf7161f5667a0d5a74487db26d3.

Commit:
Put Services into Mux nodehandle to be conform to the previous Mux-API
(as suggested by Dirk Thomas).
Corrected copyright.

dirk-thomas commented:
I will add the missing install rule (https://github.com/ros/ros_comm/pull/407/files#diff-ae340f410bae427e1853101dfdbade7eR55) when merging it into _indigo-devel_.

dirk-thomas commented:
For `mux` there are some Python scripts to ease the operations (add, delete, list, select). Wouldn't it make sense to have similar scripts for `demux`?

Squelsh commented:
I did not notice these scripts till now. Will have a look on monday.

Commit:
Added convenience scripts for the Demux.

Squelsh commented:
Wrong button )):

However: I don't see the point in a script, where the copyright is three times the lenghts of the script itself, but I committed the according scripts for the demux.

dirk-thomas commented:
Merged to indigo-devel in 4edc26b3ab23f39073d901bf1c69066c98277e81. It will be backported to hydro-devel before the next patch release. Thank you for the PR.
